### ISSUE TITLE
Can I use it for classification?

### ISSUE THREAD

**Issue by dtczhl** [facebookresearch/SparseConvNet#145]
Hi. Can I use SparseConvNet for classification? That is, given a point cloud, predicting the overall class. I can see that SparseConvNet supports scene segmentation, i.e., give a label to each point. But I would like to classify the whole point cloud rather than each point.  

Am I correct in building the following for classification (model adapted from ScanNet segmentation)? I am confused with the SparseToDense layer. 

Classify a point cloud into 100 classes. 
```
self.sparseModel = scn.Sequential().add(
           scn.InputLayer(data.dimension,data.full_scale, mode=4)).add(
           scn.SubmanifoldConvolution(data.dimension, 3, m, 3, False)).add(
               scn.UNet(data.dimension, block_reps, [m, 2*m, 3*m, 4*m, 5*m, 6*m, 7*m], residual_blocks)).add(
           scn.BatchNormReLU(m)).add(
           **scn.SparseToDense(100) ** ) .add(
           scn.OutputLayer(data.dimension))
        self.linear = nn.Linear(100, 100)
```

Thanks,

**Comment by btgraham** [comment]
Yes. You have to make sure the network is the right size for the input spatial size. Here is a very small example. You might want to replace the top convolution with average pooling.
```
class Model(nn.Module):
    def __init__(self):
        nn.Module.__init__(self)
        self.sparseModel = scn.Sequential(
          scn.SparseVggNet(3, n_input_planes, [
            ['C', 8, ], ['C', 8], 'MP',
            ['C', 16], ['C', 16], 'MP',
            ['C', 16, 8], ['C', 16, 8], 'MP',
            ['C', 24, 8], ['C', 24, 8], 'MP']),
          scn.Convolution(3, 32, 64, 5, 1, False),
          scn.BatchNormReLU(64),
          scn.SparseToDense(3, 64))
        self.spatial_size= self.sparseModel.input_spatial_size(torch.LongTensor([1, 1]))
        self.inputLayer = scn.InputLayer(3,self.spatial_size,2)
        self.linear = nn.Linear(64, 183)

    def forward(self, x):
        x = self.inputLayer(x)
        x = self.sparseModel(x)
        x = x.view(-1, 64)
        x = self.linear(x)
        return x
```

**Comment by dtczhl** [comment]
@btgraham Thanks for your reply. If I understand correctly, you apply Convolution to downgrade the input size. For example, if the spatial_size is 200, then the original input size is 200*200*200. Then after Convolutions (filter_size = 3):
1st Convolution: -> 100*100*100
2nd Convolution: -> 50*50*50
3rd Convolution: ->25*25*25
4th Convolution: -> 12*12*12   (Is it or 13*13*13 ???)
5th Convolution: -> 6*6*6
Then I can apply SparseToDense() to arrange data into 6*6*6 rows and Feature columns. 
After that, I can apply Linear to convert the data into 6*6*6 rows and 1 Feature columns as a class probability. So, In this case, I classify the 3D point cloud into 216 classes. 

So, If I want to classify a point cloud into 216 classes, I can use the following codelet
```
class Model(nn.Module):
    def __init__(self):
        nn.Module.__init__(self)
        self.sparseModel = scn.Sequential(
          scn.InputLayer(3, 200),
          scn.Convolution(3, 3, m, 3, 1, False),
          scn.BatchNormReLU(), 
          scn.Convolution(3, 3, m, 3, 1, False),
          scn.BatchNormReLU(3, m), 
         .............
          scn.Convolution(3, 3, m, 3, 1, False),
          scn.BatchNormReLU(), 
          scn.SparseToDense(3, m)
        self.linear = nn.Linear(m, 1)
    def forward(self, x):
        x = self.sparseModel(x)
        x = self.linear(x)
        return x
model=Model()
```

Could you let me know if I am correct in understanding what you said?

Thanks
